File: engine/bindings/builtin.dart
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

library sky_builtin;

import "dart:async";

// Corelib 'print' implementation.
void _print(arg) {
  _Logger._printString(arg.toString());
}

class _Logger {
  static void _printString(String s) native "Logger_PrintString";
}

class _Timer implements Timer {
  _Timer(int milliseconds,
         void callback(Timer timer),
         bool repeating) {
    _id = _create(milliseconds, () {
      if (!repeating)
        _id = 0;
      callback(this);
    }, repeating);
  }

  void cancel() {
    _cancel(_id);
    _id = 0;
  }

  bool get isActive => _id != 0;

  static int _create(int milliseconds,
                     void callback(),
                     bool repeating) native "Timer_create";
  static void _cancel(int id) native "Timer_cancel";

  int _id;
}

void _scheduleMicrotask(void callback()) native "ScheduleMicrotask";
Timer _createTimer(int milliseconds,
                   void callback(Timer timer),
                   bool repeating) {
  return new _Timer(milliseconds, callback, repeating);
}

_getPrintClosure() => _print;
_getScheduleMicrotaskClosure() => _scheduleMicrotask;
_getCreateTimerClosure() => _createTimer;

File: engine/bindings/builtin.cc
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "sky/engine/bindings/builtin.h"

#include "base/logging.h"
#include "bin/io_natives.h"
#include "dart/runtime/include/dart_api.h"
#include "gen/sky/bindings/DartGlobal.h"
#include "sky/engine/bindings/builtin_natives.h"
#include "sky/engine/bindings/builtin_sky.h"
#include "sky/engine/bindings/mojo_natives.h"
#include "sky/engine/tonic/dart_builtin.h"

namespace blink {
namespace {

struct LibraryDescriptor {
  const char* url;
  bool has_natives;
  Dart_NativeEntrySymbol native_symbol;
  Dart_NativeEntryResolver native_resolver;
};

const LibraryDescriptor kBuiltinLibraries[] = {
    /* { url_, has_natives_, native_symbol_, native_resolver_ } */
    {"dart:sky_builtin_natives",
     true,
     BuiltinNatives::NativeSymbol,
     BuiltinNatives::NativeLookup},
    {"dart:sky", true, skySnapshotSymbolizer, skySnapshotResolver},
    {"dart:mojo.internal", true, MojoNativeSymbol, MojoNativeLookup},
    {"dart:io", true, dart::bin::IONativeSymbol, dart::bin::IONativeLookup },
};

}  // namespace

void Builtin::SetNativeResolver(BuiltinLibraryId id) {
  static_assert(arraysize(kBuiltinLibraries) == kInvalidLibrary,
      "Unexpected number of builtin libraries");
  DCHECK_GE(id, kBuiltinLibrary);
  DCHECK_LT(id, kInvalidLibrary);
  if (kBuiltinLibraries[id].has_natives) {
    Dart_Handle library = DartBuiltin::LookupLibrary(kBuiltinLibraries[id].url);
    // Setup the native resolver for built in library functions.
    DART_CHECK_VALID(
        Dart_SetNativeResolver(library,
                               kBuiltinLibraries[id].native_resolver,
                               kBuiltinLibraries[id].native_symbol));
  }
}

Dart_Handle Builtin::LoadAndCheckLibrary(BuiltinLibraryId id) {
  static_assert(arraysize(kBuiltinLibraries) == kInvalidLibrary,
      "Unexpected number of builtin libraries");
  DCHECK_GE(id, kBuiltinLibrary);
  DCHECK_LT(id, kInvalidLibrary);
  Dart_Handle library = DartBuiltin::LookupLibrary(kBuiltinLibraries[id].url);
  DART_CHECK_VALID(library);
  return library;
}

}  // namespace blink
